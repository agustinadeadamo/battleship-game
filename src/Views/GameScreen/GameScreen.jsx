/**
 * @desc Dependencies
 */
import React, { useEffect, useRef, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

/**
 * @desc Actions
 */
import {
  updateWinner,
} from '../../Actions/Actions';

/**
 * @desc Components
 */
import PopUp from '../../Components/PopUp/PopUp';
import Grid from '../../Components/Grid/Grid';

/**
 * @desc Styled Components
 */
import GameScreenContainer from './style';

const StartScreen = () => {
  // State
  const [userGridGame, updateUserGrid] = useState(null);
  const [computerGridGame, updateComputerGrid] = useState(null);
  const [enableUserGrid, updateEnableUserGrid] = useState(true);
  const [shipsSelectedUser, updateShipsSelectedUser] = useState([]);
  const [shipsSelectedComputer, updateShipsSelectedComputer] = useState([]);
  const [copyPopUp, updateCopyPopUp] = useState('');
  const [showPopUp, updateShowPopUp] = useState(false);
  const [userTurn, updateUserTurn] = useState(true);

  // Store
  const userGrid = useSelector((state) => state.userGrid);
  const computerGrid = useSelector((state) => state.computerGrid);
  const user = useSelector((state) => state.user);
  const ships = useSelector((state) => state.ships);
  const dispatch = useDispatch();

  // Ref
  const prevComputerOnClickCell = useRef(0);
  const prevComputerOnClickTaken = useRef(false);
  const prevComputerOnClickCellValidfirst = useRef(0);
  const prevOptionSelected = useRef(0);

  // Router
  const history = useHistory();

  /**
   * Hook that is executed on first renders
   */
  useEffect(() => {
    // Updates state con store data
    updateUserGrid(userGrid);
    updateComputerGrid(computerGrid);
    updateShipsSelectedUser(ships);
    updateShipsSelectedComputer(ships);
  }, [userGrid, computerGrid]);

  /**
   * Function that is executed after one ship has been destroyed, validates
   * if all the ships from that user where destroyed
   *
   * @param { Array } shipsPlayer
   *
   * @return { Boolean }
   */
  const validateAllShipsWereDestroyed = (shipsPlayer) => {
    let allDestroyed = true;
    // Iterates ships
    shipsPlayer.forEach((ship) => {
      // If any of the ships have not been destroyed modificates the value to false
      if (ship.selected !== ship.spaces.length) {
        allDestroyed = false;
      }
    });

    return allDestroyed;
  };

  /**
   * Function that is executed after an onclick on user or computer grid is generated
   * Modifies the ships and the grid according to what has been found in cell selected
   *
   * @param { Array } gridToModifie
   * @param { Array } shipsToModifie
   * @param { Number } cellClickId
   * @param { Boolean } foundShip
   * @param { String } cellContent
   * @param { Boolean } cpuPlayer
   *
   * @return { Object }
   */
  const modifieGridAndShips = (gridToModifie, shipsToModifie,
    cellClickId, foundShip, cellContent, cpuPlayer = false) => {
    // Values by default
    let destroyed = false;
    let shipSelected = 0;
    let allDestroyed = false;

    const copyGridToModifie = gridToModifie;

    // Validates if a ship has been found on cell selected
    if (foundShip) {
      // Iterates ships
      shipsToModifie.forEach((ship, index) => {
        const copyShip = ship;
        // Iterates ships's spaces to know if it matchs with any content
        ship.spaces.forEach((space) => {
          if (cellContent === space) {
            shipSelected = index;
            // Updates how many cells have been selected in the ship
            copyShip.selected = ship.selected + 1;
            // Validates if all the ship cells have already been selected
            if (ship.selected === ship.spaces.length) {
              // If all the ship cells have already been selected changes status to destroyed
              copyGridToModifie[cellClickId].status = 'destroyed';
              destroyed = true;
              // Updates pop up copy
              updateCopyPopUp(`A ${shipsToModifie[shipSelected].name} has just been destroyed`);
              // If the player is the cpu updates if previous onClick has been taken
              if (cpuPlayer) {
                prevComputerOnClickTaken.current = false;
              }
            } else {
              // If not al the cells have been selected then updates status to hit
              copyGridToModifie[cellClickId].status = 'hit';
              // Updates pop up copy
              updateCopyPopUp(`A ${shipsToModifie[shipSelected].name} has just been hitten`);
              if (cpuPlayer) {
                prevComputerOnClickTaken.current = true;
              }
            }
          }
        });
      });
    } else {
      // If there is no ship in cell selected then status should be missed
      copyGridToModifie[cellClickId].status = 'missed';
      // Updates pop up copy
      updateCopyPopUp('Shot missed!');
    }

    // If there is a ship that was destroy, all the cells from that
    // ship in the grid should be updated
    if (destroyed) {
      // Iterates cells
      shipsToModifie[shipSelected].spaces.forEach((space) => {
        gridToModifie.forEach((cell) => {
          const cellCopy = cell;
          // If cell content matchs space modifies status
          if (cell.cellContent === space) {
            cellCopy.status = 'destroyed';
          }
        });
      });
      // Validate if all ships in grid have been destroyed
      allDestroyed = validateAllShipsWereDestroyed(shipsToModifie);
    }

    // Show pop up
    updateShowPopUp(true);

    // If not all have been destroyed retursn valua and ships and grid already modified
    return {
      allDestroyed,
      shipsToModifie,
      gridToModifie,
    };
  };

  /**
   * Function that generates click on computer grid
   *
   * @return { Number }
   */
  const generateComputerClick = () => {
    // Default value
    let randomClick = 0;
    // Options in case there is a logic that needs to be use when one click matchs with a ship
    const options = {
      1: prevComputerOnClickCell.current + 1,
      2: prevComputerOnClickCell.current - 1,
      3: prevComputerOnClickCell.current + 10,
      4: prevComputerOnClickCell.current - 10,
    };
    // If previous click was not a ship, generates randomclick
    if (prevComputerOnClickTaken.current === false) {
      randomClick = Math.floor(Math.random() * 99);
      // If previous click was a ship generates logic to sink the boat
      // Validates if all the options arround the previous valid click has been selected
    } else if (prevOptionSelected.current > Object.keys(options).length) {
      // If all the options arroun the preovious valid click has been
      // selected gets the first valid cell to continue with it's options
      prevComputerOnClickCell.current = prevComputerOnClickCellValidfirst.current;
      // Changes prevOptions to start from the first one
      prevOptionSelected.current = 0;
      // Generates click with first option
      randomClick = options[prevOptionSelected.current + 1];
    } else {
      // Generates click with the next option
      randomClick = options[prevOptionSelected.current + 1];
      // Updates prev options
      prevOptionSelected.current += 1;
    }

    return randomClick;
  };

  /**
   * Function that is executed every time one of the players wins the game
   * @param { String } player
   */
  const redirect = (player) => {
    // Updates in store the player who won
    dispatch(updateWinner(player));
    // Redirects to the end game screen
    history.push('/end');
  };

  /**
   * Function that is executed after each onclick in user grid, generates a
   * random click on computer grid and simulates a player actions
   */
  const onClickCellComputer = () => {
    const modifiedGridComputer = JSON.parse(JSON.stringify(userGridGame));
    const modifiedShipsComputer = JSON.parse(JSON.stringify(shipsSelectedUser));

    // Generates computer click
    const computerClick = generateComputerClick();

    // Validates if cell selected has not already been selected and is not undefined
    if (modifiedGridComputer[computerClick] !== undefined
      && modifiedGridComputer[computerClick].status !== 'hit'
      && modifiedGridComputer[computerClick].status !== 'destroyed'
      && modifiedGridComputer[computerClick].status !== 'missed') {
      // Validates if there is a ship in cell selected
      let foundShip = false;
      if (modifiedGridComputer[computerClick].cellContent !== 'sea') {
        // Updates prev computer click with ship in cell
        prevComputerOnClickCell.current = computerClick;
        foundShip = true;
        prevOptionSelected.current = 0;
      }

      // If prev click did not had a ship, updates with new cell
      if (!prevComputerOnClickTaken.current) {
        prevComputerOnClickCellValidfirst.current = computerClick;
        prevOptionSelected.current = 0;
      }

      // Updates grid an ships
      const gridAndShipsComputerModified = modifieGridAndShips(modifiedGridComputer,
        modifiedShipsComputer, computerClick, foundShip,
        modifiedGridComputer[computerClick].cellContent, true);
      // Destructurin the results
      const {
        allDestroyed = false,
        shipsToModifie = [],
        gridToModifie = [],
      } = gridAndShipsComputerModified;

      // If all ships were destroyed redirect to end game
      if (allDestroyed) {
        redirect('CPU');
      }

      // Updates grid
      updateShipsSelectedUser(shipsToModifie);
      updateUserGrid(gridToModifie);

      // Adds set time out to give time to pop up to hide
      setTimeout(() => {
        // Enables usergrid
        updateEnableUserGrid(true);
        // Modifies the turn, now it should show that it's user turn
        updateUserTurn(true);
      }, 1000);
    } else {
      // If the cell selected has already been selected then select a cell again
      onClickCellComputer();
    }
  };

  /**
   * Function that is executed when onClick in one of the cells of the userGrid
   * @param { Object } cell
   */
  const onClickCellUser = (cell) => {
    // Destructuring cell
    const {
      id = 0,
      status = 'empty',
      taken = false,
      cellContent = 'sea',
    } = cell;

    // Validates if cell has not been already selected before
    if (status !== 'hit' && status !== 'missed' && status !== 'destroyed') {
      // Validates if on that cell selected exists a ship
      let foundShip = false;
      if (taken) {
        foundShip = true;
      }

      const modifiedUserGrid = JSON.parse(JSON.stringify(computerGridGame));
      const modifiedUserShips = JSON.parse(JSON.stringify(shipsSelectedComputer));

      // Modifies grid and ship according to what have been found in cell
      const gridAndShipsUserModified = modifieGridAndShips(modifiedUserGrid,
        modifiedUserShips, id, foundShip, cellContent);
      // Destructuring the return value
      const {
        allDestroyed = false,
        shipsToModifie = [],
        gridToModifie = [],
      } = gridAndShipsUserModified;

      // If all ships where destroyed
      if (allDestroyed) {
        redirect(user);
      }

      // Updates grid and ship
      updateShipsSelectedComputer(shipsToModifie);
      updateComputerGrid(gridToModifie);
      // Disables user grid
      updateEnableUserGrid(false);
      // Updates the user turns, now it should show it's CPU turn
      updateUserTurn(false);

      // Generates a settimeout to generate a click on computer grid after 2 seconds
      setTimeout(() => { onClickCellComputer(); }, 2000);
    }
  };

  return (
    <GameScreenContainer>
      <div data-testid="popUp-container">
        <PopUp copy={copyPopUp} show={showPopUp} hidePopUp={() => { updateShowPopUp(false); }} />
      </div>
      <div className="title">
        <h1>BattleShip</h1>
        <h3>Now play!</h3>
      </div>
      <div className="player-turn">
        <p>
          Player:
          {userTurn ? user : 'CPU'}
        </p>
      </div>
      <div className="grids-container">
        <div className="first-column">
          <div data-testid="grid-container-computer" className="grid-container">
            <p className="grid-name">CPU</p>
            <Grid data-testid="grid-grid" enableGrid={false} grid={userGridGame} gridName="user" />
          </div>
        </div>
        <div className="second-column">
          <div data-testid="grid-container-user" className="grid-container">
            <p className="grid-name">{user}</p>
            <Grid onClickCell={(cell) => onClickCellUser(cell)} data-testid="grid-grid" enableGrid={enableUserGrid} grid={computerGridGame} gridName="computer" />
          </div>
        </div>
      </div>
      <div className="button-row">
        <div className="button-container">
          <button data-testid="surrender-button" type="button" onClick={() => redirect('CPU')}>
            Surrender
          </button>
        </div>
      </div>
    </GameScreenContainer>
  );
};

export default StartScreen;
