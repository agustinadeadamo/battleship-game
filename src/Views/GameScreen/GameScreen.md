# StartScreen view
This view is shows once the username was added and the ship's position has been selected. 
After each onclick in user grid it generates an onclick in computer game to simulate other player.
A pop up is shown with each notification, even if the shot has been missed.
Once all the ships where destroyed in one of the grids it redirects to the EndGame screen.