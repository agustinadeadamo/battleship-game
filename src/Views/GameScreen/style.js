/**
 * @desc Dependencies
 */
import styled from 'styled-components';

const GameScreenContainer = styled.div`
width: 90%;
max-width: 1240px;
margin: 0 auto;

  .title {
    width: 100%;
    text-align: center;
    margin: 30px 0;
  }

  .player-turn {
    text-align: center;
  }

  .grids-container {
    width: 100%;
    display: flex;
    justify-content: space-between;
    flex-wrap: wrap;

    .grid-name {
      margin: 20px 0;
      text-align: center;
    }

    .first-column, .second-column {
      width: 100%;
  
      @media (min-width: 1140px) {
        width: 45%;
      }
  
      .grid-container {
        width: 100%;
        max-width: 600px;
        margin: 20px auto;
  
        @media (min-width: 1140px) {
          margin: 20px 0;
        }
      }
    }
  }

  .button-row {
    width: 100%;

    .button-container {
      width: 30%;
      max-width: 300px;
      margin: 30px auto;
    }
  }
`;

export default GameScreenContainer;
