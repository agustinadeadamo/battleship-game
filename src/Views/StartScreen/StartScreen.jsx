/**
 * @desc Dependencies
 */
import React, { useEffect, useRef, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

/**
 * @desc Actions
 */
import {
  updateUserGridInReducer, updateComputerGridInReducer, updateUserName, updateGameStarted,
} from '../../Actions/Actions';

/**
 * @desc Components
 */
import Input from '../../Components/Input/Input';
import Grid from '../../Components/Grid/Grid';

/**
 * @desc Styled Components
 */
import StartScreenContainer from './style';

const StartScreen = () => {
  // State
  const [userGrid, updateUserGrid] = useState(null);
  const [computerGrid, updateComputerGrid] = useState(null);
  const [showMessageErrorUser, updateShowMessageErrorUser] = useState(false);

  // Router
  const history = useHistory();

  // Store
  const gridWidth = useSelector((state) => state.gridWidth);
  const gridHeight = useSelector((state) => state.gridHeight);
  const user = useSelector((state) => state.user);
  const ships = useSelector((state) => state.ships);
  const dispatch = useDispatch();

  // Refs
  const prevUserGridRef = useRef([]);
  const prevComputerGridRef = useRef([]);

  /**
   * Function that generates each ship in the grid
   * @param { Object } ship
   * @param { Array } grid
   *
   * @return { Array }
   */
  const generateShip = (ship, grid) => {
    // Validates that gris is an array
    if (Array.isArray(grid)) {
      // Defines direction by default
      let direction = 1;
      // Choose horizontaly or vertically direction randomly
      const randomDirection = Math.floor(Math.random() * ship.directions.length);
      // According to the random direction selected, choose the ship with the specified direciton
      const current = ship.directions[randomDirection];

      if (randomDirection === 0) {
        // This determinates the spaces between each cell according to the direction selected
        direction = 1;
      }
      if (randomDirection === 1) {
        // This determinates the spaces between each cell according to the direction selected
        direction = 10;
      }

      // Defines the las cell available according to the position selected
      const lastCellAvailable = grid.length - (ship.directions[0].length * direction);
      // Defines the random start of the ship
      const randomStart = Math.floor(Math.random() * lastCellAvailable);
      // Defines if the cell is taken
      const isTaken = current.some((index) => grid[randomStart + index].taken);
      // Defines if the random start has enough spaces on the right
      const isAtRightEdge = current.some((index) => (randomStart + index) % 10 === 10 - 1);
      // Defines if the random start has enough spaces on the right
      const isAtLeftEdge = current.some((index) => (randomStart + index) % 10 === 0);

      const modifiedInitialGrid = JSON.parse(JSON.stringify(grid));

      if (!isTaken && !isAtRightEdge && !isAtLeftEdge) {
        // Creates the ship in the grid
        current.forEach((value, index) => {
          modifiedInitialGrid[randomStart + value].taken = true;
          modifiedInitialGrid[randomStart + value].status = 'taken';
          modifiedInitialGrid[randomStart + value].cellContent = ship.spaces[index];
        });

        return modifiedInitialGrid;
      }

      // If it's taken or does not has enough spaces on the right and on the left
      // ship will be generated again
      return generateShip(ship, grid);
    }

    return [];
  };

  /**
   * Function that generates an initial and empty grid
   *
   * @return { Array }
   */
  const generateInitialGrid = () => {
    const initialGrid = [];
    // Creates grid cells with it's properties
    for (let i = 0; i < gridWidth * gridHeight; i += 1) {
      initialGrid.push({
        id: i,
        status: 'empty',
        taken: false,
        cellContent: 'sea',
      });
    }

    return initialGrid;
  };

  /**
   * Hook that is executed in first mount
   */
  useEffect(() => {
    // Generates initial grid
    const initialUserGrid = generateInitialGrid();

    // Updates userGrid and computerGrid
    updateUserGrid(initialUserGrid);
    updateComputerGrid(initialUserGrid);
  }, []);

  /**
   * Function that iterates the ship to generate each one
   * @param {*} gridForShips
   *
   * @return { Array }
   */
  const iterateShips = (gridForShips) => {
    let newGridForShips = gridForShips;
    ships.forEach((ship) => {
      newGridForShips = generateShip(ship, newGridForShips);
    });
    return newGridForShips;
  };

  /**
   * Hook that is executed on first render and every time userGrid or ships update
   */
  useEffect(() => {
    // Validates if userGrid is an array and has content, if ships is an array and has
    // content and if the prev value for userGrid was an empty array
    if ((Array.isArray(userGrid) && userGrid.length > 0
    && Array.isArray(ships) && ships.length > 0
    && prevUserGridRef.current.length === 0)) {
      // Updates prev value of userGrid
      prevUserGridRef.current = userGrid;
      // Iterates ship
      const userGridWithShips = iterateShips(userGrid);
      // Updates state
      updateUserGrid(userGridWithShips);
    }
  }, [userGrid, ships]);

  /**
   * Hook that is executed on first render and every time computerGrid and ships update
   */
  useEffect(() => {
    // Validates if computerGrid is an array and has content, if ships is an array and has
    // content and if the prev value for computerGrid was an empty array
    if ((Array.isArray(computerGrid) && computerGrid.length > 0
    && Array.isArray(ships) && ships.length > 0
    && prevComputerGridRef.current.length === 0)) {
      // Updates prev value of computerGrid
      prevComputerGridRef.current = computerGrid;
      // Iterates ship
      const computerGridWithShips = iterateShips(computerGrid);
      // Updates state
      updateComputerGrid(computerGridWithShips);
    }
  }, [computerGrid, ships]);

  /**
   * Function that is executed when "Start game" button is selected
   */
  const onClickStartGame = () => {
    let validated = true;

    // Validates if there is a user added
    if (user === '') {
      validated = false;
      updateShowMessageErrorUser(true);
    }

    // If user is added then updates store and redirects
    if (validated) {
      dispatch(updateUserName(user));
      dispatch(updateUserGridInReducer(userGrid));
      dispatch(updateComputerGridInReducer(computerGrid));
      dispatch(updateGameStarted(true));
      history.push('/game');
    }
  };

  /**
   * Function that is executed when "position ships" is clicked
   */
  const onClickChangePositions = () => {
    // Updates prev value for user Grid
    prevUserGridRef.current = [];
    // Generates initial grid
    const initialUserGrid = generateInitialGrid();
    // Updates state
    updateUserGrid(initialUserGrid);
  };

  /**
   * Function that is executed when input changes value
   */
  const onChangeUserNameInput = (value) => {
    // If value is different to an empty string it hides the error message
    if (value !== '') {
      updateShowMessageErrorUser(false);
    }

    // Updates store
    dispatch(updateUserName(value));
  };

  return (
    <StartScreenContainer>
      <div className="title">
        <h1>Battleship</h1>
        <h3>Configuration</h3>
      </div>
      <div className="instructions">
        <ul>
          <li>Enter your user name.</li>
          <li>Place you ships on the board the way you prefer.</li>
          <li>Start the game !</li>
        </ul>
      </div>
      <div className="configuration">
        <div className="first-column">
          <div data-testid="grid-container" className="grid-container">
            <Grid data-testid="grid-grid" enableGrid={false} grid={userGrid} gridName="user" />
          </div>
        </div>
        <div className="second-column" data-testid="actions-container">
          <div className="actions">
            <Input messageError="You must add a user name" showMessageError={showMessageErrorUser} data-testid="username-input" placeholder="Insert user name here" onChangeValue={(valueInput) => onChangeUserNameInput(valueInput)} value={user} />
            <div className="buttons-container">
              <button data-testid="change-positions-button" type="button" onClick={() => onClickChangePositions()}> Change position </button>
              <button className="dark" type="button" data-testid="start-game" onClick={() => onClickStartGame()}>Start Game</button>
            </div>
          </div>
        </div>
      </div>
    </StartScreenContainer>
  );
};

export default StartScreen;
