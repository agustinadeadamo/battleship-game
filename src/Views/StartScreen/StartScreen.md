# StartScreen view
This view is when at the beggining of the game. 
Makes the user to choose a ship's position and enter user name.
If the user name is empty it shows a message error.
If the user name is filled it redirects to the GameScreen where the user and computer grid will be shown.
This view also generates the computer grid once the gamer clicks on "Start game" button.