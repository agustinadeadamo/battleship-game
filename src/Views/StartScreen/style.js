/**
 * @desc Dependencies
 */
import styled from 'styled-components';

const StartScreenContainer = styled.div`
  width: 100%;

  .title, .instructions {
    width: 100%;
    text-align: center;
    margin: 30px 0;
  }

  .configuration {
    width: 90%;
    max-width: 1240px;
    margin: 0 auto;
    display: flex;
    justify-content: space-arround;
    flex-wrap: wrap;

    .first-column {
      width: 100%;
  
      @media (min-width: 1140px) {
        width: 70%;
      }
  
      .grid-container {
        width: 100%;
        max-width: 600px;
        margin: 20px auto;
  
        @media (min-width: 1140px) {
          margin: 0;
        }
      }
    }
  
    .second-column {
      width: 100%;
  
      @media (min-width: 1140px) {
        width: 30%;
      }
  
      .actions {
        width: 100%;
        max-width: 600px;
        margin: 20px auto;
  
        @media (min-width: 1140px) {
          margin: 0;
        }

        .buttons-container {
          margin: 30px;
        }
      }
    }
  }
`;

export default StartScreenContainer;
