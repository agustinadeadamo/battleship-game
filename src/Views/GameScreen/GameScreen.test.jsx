/**
 * @desc Dependencies
 */
import React from 'react';
import { render, fireEvent, within } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import {
  describe, expect, it, beforeEach,
} from '@jest/globals';
import { createMemoryHistory } from 'history';
import { act, create } from 'react-test-renderer';

/**
 * @desc Store
 */
import Store from '../../Config/Store';

/**
 * @desc Component
 */
import GameScreen from './GameScreen';
import StartScreen from '../StartScreen/StartScreen';

describe('[GameScreen View]', () => {
  let getByTestId;
  const history = createMemoryHistory();

  beforeEach(() => {
    ({ getByTestId } = render(
      <Router history={history}>
        <Provider store={Store}>
          <StartScreen />
          <GameScreen />
        </Provider>
      </Router>,
    ));

    // Gets start game button
    const startGameButton = getByTestId('start-game');
    // Gets input
    const actions = getByTestId('actions-container');
    const input = within(actions).getByTestId('input-component');
    // Fires onChange event in input
    act(() => { fireEvent.change(input, { target: { value: 'text' } }); });
    // Fires onClick event start game button
    act(() => { fireEvent.click(startGameButton); });
  });

  it('Matchs snapshot', () => {
    const startscreen = create(
      <Provider store={Store}>
        <GameScreen />
      </Provider>,
    );
    expect(startscreen).toMatchSnapshot();
  });

  it('Tests if the grid was shown correctly in user grid', () => {
    // Get grid container
    const gridContainer = getByTestId('grid-container-user');
    // Gets all cells
    const cells = within(gridContainer).getAllByTestId(/cell-component/);
    // Gets each ship
    const cruiser = within(gridContainer).getAllByTestId(/cruiser/);
    const submarine = within(gridContainer).getAllByTestId(/submarine/);
    const carrier = within(gridContainer).getAllByTestId(/carrier/);

    expect(cells).toHaveLength(100);
    expect(cruiser).toHaveLength(9);
    expect(submarine).toHaveLength(2);
    expect(carrier).toHaveLength(4);
    expect(gridContainer).toBeInTheDocument();
  });

  it('Tests if the grid was shown correctly in computer grid', () => {
    // Get grid container
    const gridContainer = getByTestId('grid-container-computer');
    // Gets all cells
    const cells = within(gridContainer).getAllByTestId(/cell-component/);
    // Gets each ship
    const cruiser = within(gridContainer).getAllByTestId(/cruiser/);
    const submarine = within(gridContainer).getAllByTestId(/submarine/);
    const carrier = within(gridContainer).getAllByTestId(/carrier/);

    expect(cells).toHaveLength(100);
    expect(cruiser).toHaveLength(9);
    expect(submarine).toHaveLength(2);
    expect(carrier).toHaveLength(4);
    expect(gridContainer).toBeInTheDocument();
  });

  it('Tests if when ship is hit shows pop up with right message', () => {
    // Get grid container
    const gridContainer = getByTestId('grid-container-user');
    // Get first cell o first cruiser
    const cruiser = within(gridContainer).getByTestId(/cruisers-1-1/);
    // Fire event onclick in the cell
    act(() => { fireEvent.click(cruiser); });
    // Gets popUp message
    const popUpContainer = getByTestId('popUp-container');
    const popUpMessage = within(popUpContainer).getByText('A cruisers has just been hitten');

    expect(popUpMessage).toBeInTheDocument();
  });

  it('Tests if when shot is missed shows pop up with right message', () => {
    // Get grid container
    const gridContainer = getByTestId('grid-container-user');
    // Get cell with sea content
    const sea = within(gridContainer).getAllByTestId(/sea/);
    // Fire event onclick in the cell
    act(() => {fireEvent.click(sea[0])});
    // Gets popUp message
    const popUpContainer = getByTestId('popUp-container');
    const popUpMessage = within(popUpContainer).getByText('Shot missed!');

    expect(popUpMessage).toBeInTheDocument();
  });
});
