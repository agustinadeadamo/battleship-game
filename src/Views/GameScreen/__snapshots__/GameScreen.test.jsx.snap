// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GameScreen View] Matchs snapshot 1`] = `
<div
  className="sc-dlfnbm jjBisn"
>
  <div
    data-testid="popUp-container"
  >
    <div
      className="sc-bdfBwQ hWUWEm hide"
      data-testid="popUp-component"
    >
      
    </div>
  </div>
  <div
    className="title"
  >
    <h1>
      BattleShip
    </h1>
    <h3>
      Now play!
    </h3>
  </div>
  <div
    className="player-turn"
  >
    <p>
      Player:
      text
    </p>
  </div>
  <div
    className="grids-container"
  >
    <div
      className="first-column"
    >
      <div
        className="grid-container"
        data-testid="grid-container-computer"
      >
        <p
          className="grid-name"
        >
          CPU
        </p>
        <div>
          <div
            className="sc-gsTCUz fzmsNw grid-user"
            data-testid="grid-component"
          >
            <div
              className="content"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="second-column"
    >
      <div
        className="grid-container"
        data-testid="grid-container-user"
      >
        <p
          className="grid-name"
        >
          text
        </p>
        <div>
          <div
            className="sc-gsTCUz bIPugA grid-computer"
            data-testid="grid-component"
          >
            <div
              className="content"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="button-row"
  >
    <div
      className="button-container"
    >
      <button
        data-testid="surrender-button"
        onClick={[Function]}
        type="button"
      >
        Surrender
      </button>
    </div>
  </div>
</div>
`;
